content/en/blog/attribute-autocomplete-enhancement/index.md

```diff
@@ -0,0 +1,306 @@
+description: ""
+lead: "We discuss the two new enhancement added to the attribute name autocomplete feature in Texera, auto attribute correction and attribute type suggestion"
+date: 2023-06-18T20:58:20-07:00
+lastmod: 2023-06-18T20:58:20-07:00
+draft: true
+weight: 50
+#images: ["attribute-autocomplete-enhancement.jpg"]
+contributors: ["Chengxi Li"]
+---
+
+# Enhancing Attribute Autocomplete Functionality -  Auto Correction and Type Suggestion
+
+In this blog, we discuss the two new enhancement added to the attribute name autocomplete feature in Texera, auto attribute correction and attribute type suggestion, making the feature more user-friendly.
+
+## Background
+
+### What are operator properties
+
+Texera workflows consist of sequences of operators. Every operator has properties as the parameters of the data process function. For example, in Fig. 1, the "Filter" operator has three properties, `Attribute`, `Condition`, and `Value`. The property values in Fig. 1 indicate that the operator filters the records with amount greater than 1.
+
+**![](https://lh6.googleusercontent.com/lxGfMeDfnw14f4NskeiSVBlz3_CFUQYrQJ8l97GnDo5L5LXoZ4w8mxJUa9AnnOWn7FLlxO-kwmValePube-Hw21BlwuhzFLco_pPhdrJu7SHObcQXJEtug4h-oBDIqnXdGXfwYlerpGvkZyaZ84bK4s)**
+
+Fig. 1, Filter operator properties
+
+### What is operator input schema
+
+Various formats of data can be processed by Texera, and the format of data is represented by its schema, which is a set of attributes. In order to let every operator know what attributes of the data are available, each operator deduces its output schema using its input schema and propagates it to the next operator. This mechanism is called "schema propagation." Schema propagation is an important part of the system, as it is the base of many useful features such as operator validation and attribute autocomplete.
+
+### What is attribute autocomplete
+
+Many operators have properties that take an attribute name of the input data as value. For example, the `Attribute` property in the "Sentiment Analysis" operator takes the name of the attribute on which it will perform analysis. Autocomplete feature allows users to select an input attribute from a dropdown menu instead of manually typing in the name.
+
+**![](https://lh3.googleusercontent.com/vqgcIM-ae7mRCIbah-ZyZRJkM4T4MMNdo3HGQtVT0RFfeh9J69z7SIUspMfY1IUjJ0V30U84PLdBMd3QouXKLEbfjUBc3Ac5q3_CtKk5XAP55I3QY0oG2U6Q11TIWcXUrVfjhbGpUUgV7LPp1R4ACeM)**
+
+Fig. 2, Dropdown menu for attribute autocomplete
+
+## Auto Attribute Correction
+
+### Motivation
+
+In the previous autocomplete implementation, when an attribute name in the input schema of an operator changes, the field of this operator that selects the attribute will not update its value according to the change in the attribute name, and the operator will therefore become invalid. Intuitively, it is expected that the field will be updated accordingly. 
```

Comment:
what do you mean by previous? be more specific. maybe describe the old design in the backgrounds as well.

Reply:
There is no design to handle this in the past. As described here, the system won't do anything when an attribute is renamed or deleted. 

I am happy to explain it more clear, but I don't know how I should explain the old design because there was not design.

Reply:
we had schema propagation logic before. It works by taking a topological order of the DAG and propagate the schema from source operators all the way to sink operators. I think it's worth to explain the original schema propagation logic before you dive into the motivation.